File: Java/Unit 02/Static/Q4.java
Single File Programming Question
Problem Statement



Emily runs a bakery and tracks the monthly profits and losses of her business over a period of 
  time. She wants to calculate her maximum net profit by removing one continuous subarray (one or more consecutive months) 
  with the highest loss (i.e., the subarray with the smallest sum).



However:

If all months have only one loss, no subarray is removed.
Otherwise, she removes the continuous subarray with the most significant loss, and the rest of the months contribute to the net profit.


Your task is to help Emily determine the maximum possible net profit she can achieve after following these rules.



Example



Input:

6

1000 -200 300 -500 400 300

Output:

1800

Explanation:

The month in which the business performed poorly was with a loss of 500. So, after ignoring that month, the net profit is 1000-200+300+400+300 = 1800.

Input format :
The first line of input consists of an integer N, representing the number of monthly profit/loss records.

The second line consists of N space-separated integers, each representing the profit or loss for each month (Negative values represent losses).

Output format :
The output prints a single integer representing the net profit.



Refer to the sample output for formatting specifications.

Code constraints :
The given testcases fall under the following constraints:

1 ≤ N ≤ 10

-100000 ≤ array elements ≤ 100000

Sample test cases :
Input 1 :
4
500 -200 100 300
Output 1 :
900
Input 2 :
6
1000 -200 300 -500 400 300
Output 2 :
1800



  import java.util.Scanner;

public class Main {
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();

        int[] arr = new int[N];

        for (int i = 0; i < N; i++) {
            arr[i] = sc.nextInt();
        }

        int totalSum = 0;
        for (int i = 0; i < N; i++) {
            totalSum += arr[i];
        }

        // Find minimum subarray sum (Kadane’s algorithm for min sum)
        int minEndingHere = arr[0];
        int minSoFar = arr[0];

        for (int i = 1; i < N; i++) {
            minEndingHere = Math.min(arr[i], minEndingHere + arr[i]);
            minSoFar = Math.min(minSoFar, minEndingHere);
        }

        // If removing improves total, subtract minSoFar
        if (minSoFar < 0) {
            System.out.println(totalSum - minSoFar);
        } else {
            System.out.println(totalSum);
        }

        sc.close();
    }
}

File: DSA /Unit 05/Lec 30/Q4.cpp
Single File Programming Question
Problem Statement



Kavya is organizing item codes in her database and wants them sorted first before separating them into odd and even categories. She uses Heap Sort to arrange all numbers in ascending order, then lists all odd numbers first and even numbers next. 



Help her generate the required output.

Input format :
The first line contains an integer n, representing the number of item codes.

The second line contains n space-separated integers, representing the item codes.

Output format :
The first line of output prints all odd numbers in ascending order.

The second line prints all even numbers in ascending order.



Refer to the sample output for formatting specifications.

Code constraints :
The given test cases fall under the following constraints:

1 ≤ n ≤ 100

1 ≤ item code ≤ 105

Sample test cases :
Input 1 :
5
9 4 2 7 3
Output 1 :
3 7 9 
2 4 
Input 2 :
6
10 21 3 14 7 8
Output 2 :
3 7 21 
8 10 14

#include <iostream>
#include <algorithm>
using namespace std;

int main() {
    int n;
    cin >> n;

    int a[100];
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    sort(a, a + n);

    for (int i = 0; i < n; i++) {
        if (a[i] % 2 != 0) {
            cout << a[i] << " ";
        }
    }
    cout << endl;

    for (int i = 0; i < n; i++) {
        if (a[i] % 2 == 0) {
            cout << a[i] << " ";
        }
    }

    return 0;
}
